File: python/02_selection_sort/selection_sort.py
class SelectionSort:

    def selection_sort(self, arr):
        output_arr = []
        arr_copy = list(arr)  # make a copy of the array so original is not mutated

        for i in range(len(arr_copy)):
            smallest_index = self.find_smallest(arr_copy)
            output_arr.append(arr_copy.pop(smallest_index))

        return output_arr

    def find_smallest(self, arr):
        smallest_index = 0
        smallest_number = arr[0]

        for i in range(1, len(arr)):
            if arr[i] < smallest_number:
                smallest_index = i
                smallest_number = arr[i]

        return smallest_index


if __name__ == "__main__":
    ss = SelectionSort()
    my_list = [5, 3, 6, 2, 10, -5]

    print(ss.selection_sort(my_list))

File: java/src/main/java/ch2_selection_sort/SelectionSort.java
package ch2_selection_sort;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SelectionSort {
    public static void main(String[] args) {
        List<Integer> testList = new ArrayList<>(Arrays.asList(5, 3, 6, 2, 10, -5));
        System.out.println(selection_sort(testList));
    }

    public static List<Integer> selection_sort(List<Integer> arr) {
        List<Integer> outputList = new ArrayList<>(arr.size());
        List<Integer> copyList = new ArrayList<>(arr); // Copy list so original isn't mutated

        int originalSize = copyList.size();
        for (int i = 0; i < originalSize; i++) {
            int smallest_index = findSmallest(copyList);
            outputList.add(copyList.get(smallest_index));
            copyList.remove(smallest_index);
        }
        return outputList;
    }

    public static int findSmallest(List<Integer> arr) {
        int smallestNumber = arr.getFirst();
        int smallestIndex = 0;

        for (int i = 0; i < arr.size(); i++) {
            if( arr.get(i) < smallestNumber) {
                smallestNumber = arr.get(i);
                smallestIndex = i;
            }
        }
        return smallestIndex;
    }
}
